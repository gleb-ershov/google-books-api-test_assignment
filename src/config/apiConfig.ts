export const _apikey = "AIzaSyCvrUoNbctrhVFjh-Q1i8g8vmIK0luDkQk";
export const _baseUrl = "https://www.googleapis.com/books/v1/volumees";
