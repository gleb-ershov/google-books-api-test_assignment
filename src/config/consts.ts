import { IBookData } from "src/types";

export const FORM_SORT_OPTIONS = ["Relevance", "Newest"];

export const VOLUME_MOCK_DATA = {
	kind: "books#volume",
	id: "z7D2DwAAQBAJ",
	etag: "wvi0tv9wXHw",
	selfLink: "https://www.googleapis.com/books/v1/volumes/z7D2DwAAQBAJ",
	volumeInfo: {
		title: "The Modern JavaScript Basics Tutorial",
		subtitle:
			"JavaScript from A to Z : Learn The JavaScript Programming Language Completely From Scratch",
		authors: ["Be Sure Academy"],
		publisher: "Sure Academy",
		publishedDate: "2023-01-15",
		description:
			"JavaScript is the world's most popular programming language. JavaScript is the programming language of the Web. JavaScript is easy to learn. Here we learn JavaScript, starting from scratch and go on to advanced concepts like OOP. You will learn by examples In this book, learning speed is your choice. Everything is up to you. If you are struggling, take a break, or re-read the material. JavaScript is already running in your browser on your computer, on your tablet, and on your smart-phone. JavaScript is free to use for everyone. Start your coding today!",
		readingModes: {
			text: false,
			image: true,
		},
		pageCount: 78,
		printType: "BOOK",
		categories: ["Computers"],
		maturityRating: "NOT_MATURE",
		allowAnonLogging: true,
		contentVersion: "2.2.3.0.preview.1",
		panelizationSummary: {
			containsEpubBubbles: false,
			containsImageBubbles: false,
		},
		imageLinks: {
			smallThumbnail:
				"http://books.google.com/books/content?id=z7D2DwAAQBAJ&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api",
			thumbnail:
				"http://books.google.com/books/content?id=z7D2DwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api",
		},
		language: "en",
		previewLink:
			"http://books.google.com/books?id=z7D2DwAAQBAJ&printsec=frontcover&dq=js&hl=&as_pt=BOOKS&cd=1&source=gbs_api",
		infoLink:
			"http://books.google.com/books?id=z7D2DwAAQBAJ&dq=js&hl=&as_pt=BOOKS&source=gbs_api",
		canonicalVolumeLink:
			"https://books.google.com/books/about/The_Modern_JavaScript_Basics_Tutorial.html?hl=&id=z7D2DwAAQBAJ",
	},
	saleInfo: {
		country: "GE",
		saleability: "NOT_FOR_SALE",
		isEbook: false,
	},
	accessInfo: {
		country: "GE",
		viewability: "PARTIAL",
		embeddable: true,
		publicDomain: false,
		textToSpeechPermission: "ALLOWED",
		epub: {
			isAvailable: false,
		},
		pdf: {
			isAvailable: true,
			acsTokenLink:
				"http://books.google.com/books/download/The_Modern_JavaScript_Basics_Tutorial-sample-pdf.acsm?id=z7D2DwAAQBAJ&format=pdf&output=acs4_fulfillment_token&dl_type=sample&source=gbs_api",
		},
		webReaderLink:
			"http://play.google.com/books/reader?id=z7D2DwAAQBAJ&hl=&as_pt=BOOKS&source=gbs_api",
		accessViewStatus: "SAMPLE",
		quoteSharingAllowed: false,
	},
	searchInfo: {
		textSnippet:
			"JavaScript is the programming language of the Web. JavaScript is easy to learn. Here we learn JavaScript, starting from scratch and go on to advanced concepts like OOP. You will learn by examples In this book, learning speed is your choice.",
	},
};

export const MOCK_STATE = {
	booksData: {
		data: { items: [VOLUME_MOCK_DATA], totalItems: 1 },
		loadingStatus: "idle" as "idle" | "pending",
		error: null,
		nothingFound: false,
		currentRequestId: "",
	},
	singleBookData: {
		data: null as null | IBookData,
		loadingStatus: "idle" as "idle" | "pending",
		error: null,
	},
};
