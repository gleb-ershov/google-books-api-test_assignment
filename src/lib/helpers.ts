export const bookImageAltConstructor = (title?: string, authors?: string[]) => {
	return `${title} ${authors ? `by ${authors.join(", ")} ` : ""}volume cover`;
};

export const bookImageSrcConstructor = (
	src?: string,
	width: string | number = 150,
	height: string | number = 230
) => {
	return src || `https://placehold.co/${width}x${height}}`;
};
