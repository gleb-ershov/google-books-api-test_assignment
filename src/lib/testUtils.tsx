import React, { PropsWithChildren, ReactElement, useState } from "react";

import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "src/store";

import { BrowserRouter, MemoryRouter } from "react-router-dom";

import userEvent from "@testing-library/user-event";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";

interface IExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
	preloadedState?: Partial<RootState>;
	store?: AppStore;
}

interface IExtendedRenderOptionsWithRouter extends IExtendedRenderOptions {
	routerType?: string;
	initialEntries?: string[];
	initialIndex?: number;
}

export function renderWithProviders(
	ui: React.ReactElement,
	extendedRenderOptions: IExtendedRenderOptionsWithRouter = {}
) {
	const {
		preloadedState = {},
		routerType = "browser",
		initialEntries,
		initialIndex,
		store = setupStore(preloadedState),
		...renderOptions
	} = extendedRenderOptions;

	interface IWrapper extends PropsWithChildren {
		locationValue?: string[];
	}

	const Wrapper = ({ children }: IWrapper) => (
		<Provider store={store}>
			{routerType === "browser" ? (
				<BrowserRouter>{children}</BrowserRouter>
			) : (
				<MemoryRouter initialEntries={initialEntries}>
					{children}
				</MemoryRouter>
			)}
		</Provider>
	);

	return {
		store,
		...render(ui, { wrapper: Wrapper, ...renderOptions }),
	};
}

interface IControllingComponent extends PropsWithChildren {
	defaultValue: string;
	mockOnChange: jest.Mock<void, [str: string]>;
}

export const ControllingComponent = (props: IControllingComponent) => {
	const { defaultValue, mockOnChange, children } = props;
	const [value, setValue] = useState(defaultValue);
	mockOnChange.mockImplementation((v) => setValue(v));

	return (
		<>
			{children
				? React.cloneElement(children as ReactElement, {
						value,
						handler: mockOnChange,
				  })
				: null}
		</>
	);
};

export const setupTestWithUserEvent = () => {
	const user = userEvent.setup();
	let value = "";
	const handleEvent = jest.fn();

	return { user, handleEvent, value };
};
