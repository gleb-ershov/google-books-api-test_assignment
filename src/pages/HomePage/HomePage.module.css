.HomePage_loadMoreButton_container {
	margin: 0 auto 3em auto;
	padding: 0;
	width: 150px;
}
