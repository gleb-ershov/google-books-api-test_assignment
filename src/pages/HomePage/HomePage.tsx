import { useEffect, useMemo } from "react";

import { useAppDispatch, useAppSelector } from "src/store";
import { getBooksData } from "src/store/slices/dataSlice";
import { updateStartIndexValue } from "src/store/slices/searchFormSlice";

import { useToaster } from "src/hooks";

import {
	SearchForm,
	BooksTable,
	Toaster,
	NotFoundNotification,
} from "src/components/features";
import { CustomButton, LoaderSpinner } from "src/components/ui";

import styles from "./HomePage.module.css";

const HomePage = () => {
	const { startIndex: currentIndex } = useAppSelector(
		(state) => state.searchFormReducer
	);

	const {
		loadingStatus,
		data: { items },
		nothingFound,
		error,
	} = useAppSelector((state) => state.dataReducer.booksData);

	const memoizedError = useMemo(() => error, [error]);
	const dispatch = useAppDispatch();

	const loadMoreHandler = () => {
		dispatch(updateStartIndexValue(currentIndex + 30));
		dispatch(getBooksData());
	};

	const { showToaster, toasterProps, isToasterVisible } = useToaster();

	useEffect(() => {
		if (memoizedError) {
			showToaster.error(memoizedError.message, memoizedError.name);
		}
	}, [memoizedError, showToaster]);

	const isPending = loadingStatus === "pending";

	const notFoundNotification = nothingFound && !isPending && (
		<NotFoundNotification />
	);

	const booksTable =
		(!isPending && items) || (isPending && items && currentIndex > 0) ? (
			<BooksTable />
		) : null;

	const loadingIndicator = isPending && <LoaderSpinner />;

	const toaster = isToasterVisible && <Toaster {...toasterProps} />;

	const loadMoreButton =
		items.length && !isPending ? (
			<CustomButton
				handler={() => loadMoreHandler()}
				content="Load more"
			/>
		) : null;

	const loadMoreLoadingIndicator =
		items.length && isPending && currentIndex ? <LoaderSpinner /> : null;

	return (
		<>
			{toaster}
			<SearchForm />
			{notFoundNotification}
			{loadingIndicator}
			{booksTable}
			<div className={styles.HomePage_loadMoreButton_container}>
				{loadMoreLoadingIndicator}
				{loadMoreButton}
			</div>
		</>
	);
};

export default HomePage;
