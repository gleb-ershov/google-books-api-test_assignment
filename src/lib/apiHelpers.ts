import { _apikey, _baseUrl } from "src/config/apiConfig";

export class ResponseError extends Error {
	statusCode = 0;
	info = "";

	constructor(message: string, statusCode: number) {
		super(message);
		this.name = "ResponseError";
		this.statusCode = statusCode;
	}
}

export const urlConstructor = (
	startIndex: number,
	searchQuery: string,
	sortQuery: string = "Relevance"
) => {
	const params = [{ q: searchQuery }, { orderBy: sortQuery }, { startIndex }];
	let _url = `${_baseUrl}?`;
	params.forEach((item) => {
		const key = Object.keys(item)[0];
		_url += key + "=" + item[key as keyof typeof item] + "&";
	});
	_url += `maxResults=30&printType=books&key=${_apikey}`;
	return _url;
};

export const getBooksByQuery = async (
	startIndex: number,
	searchQuery: string,
	sortQuery: string = "Relevance"
) => {
	const requestUrl = urlConstructor(startIndex, searchQuery, sortQuery);

	try {
		const response = await fetch(requestUrl);

		if (!response.ok) {
			throw new ResponseError(
				"An error occured in response",
				response.status
			);
		}
		return await response.json();
	} catch (error) {
		if (error instanceof ResponseError || error instanceof Error) {
			throw error;
		}
		throw new Error("Uncaught error");
	}
};

export const getBooksById = async (bookId: string) => {
	try {
		const response = await fetch(_baseUrl + `/${bookId}`);

		if (!response.ok) {
			throw new ResponseError(
				"An error occured in response",
				response.status
			);
		}
		return await response.json();
	} catch (error) {
		if (error instanceof ResponseError || error instanceof Error) {
			throw error;
		}
		throw new Error("Uncaught error");
	}
};
