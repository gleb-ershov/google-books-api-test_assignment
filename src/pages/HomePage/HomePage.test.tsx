import { MOCK_STATE } from "src/config/consts";
import HomePage from "./HomePage";
import { renderWithProviders } from "src/lib/testUtils";
import { screen } from "@testing-library/react";

describe("Home Page", () => {
	it("should renders correctly if loading status is idle and have data", async () => {
		const { container } = renderWithProviders(<HomePage />, {
			preloadedState: {
				dataReducer: MOCK_STATE,
			},
		});
		expect(screen.getByRole("img")).toBeInTheDocument();
		expect(screen.getByRole("img")).toHaveAttribute(
			"alt",
			"The Modern JavaScript Basics Tutorial by Be Sure Academy volume cover"
		);
		expect(container).toMatchSnapshot();
	});

	it("should render load more button with correct text content", async () => {
		renderWithProviders(<HomePage />, {
			preloadedState: {
				dataReducer: MOCK_STATE,
			},
		});
		expect(screen.getByText("Load more")).toBeInTheDocument();
	});

	it("should render loader instead of load more button when pending", async () => {
		const newState = { ...MOCK_STATE };
		newState.booksData = { ...MOCK_STATE.booksData };
		newState.booksData.loadingStatus = "pending";

		renderWithProviders(<HomePage />, {
			preloadedState: {
				dataReducer: newState,
			},
		});
		const buttons = screen.getAllByRole("button");
		const loaders = screen.getAllByRole("alert");

		expect(loaders.length).toBe(1);
		expect(buttons.length).toBe(1);
		expect(screen.getByRole("button")).not.toHaveTextContent("Load more");
	});

	it("should render `Nothing found` notification instead of content table", async () => {
		const newState = { ...MOCK_STATE };
		newState.booksData = { ...MOCK_STATE.booksData };
		newState.booksData.nothingFound = true;
		newState.booksData.data = { ...MOCK_STATE.booksData.data };
		newState.booksData.data.items = [];

		renderWithProviders(<HomePage />, {
			preloadedState: {
				dataReducer: newState,
			},
		});
		const buttons = screen.getAllByRole("button");
		expect(buttons.length).toBe(1);
		expect(screen.getByRole("alert")).toBeInTheDocument();
		expect(screen.getByRole("alert")).toHaveTextContent(
			"Nothing was found"
		);
	});
});
