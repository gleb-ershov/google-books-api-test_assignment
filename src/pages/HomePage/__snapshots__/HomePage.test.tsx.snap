// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page should renders correctly if loading status is idle and have data 1`] = `
<div>
  <form
    class="SearchForm"
    role="form"
  >
    <h1
      class="SearchForm_header"
    >
      Search a book
    </h1>
    <div
      class="SearchForm_input_container"
    >
      <label
        class="SearchForm_input_label
     
      "
        for="searchForm_searchbar"
      >
        Search
      </label>
      <input
        class="CustomInput focused_searchbar "
        id="searchForm_searchbar"
        placeholder=""
        required=""
        role="search"
        type="search"
        value=""
      />
    </div>
    <div
      class="SearchForm_select_container"
    >
      <label
        class="SearchForm_label"
        for="searchForm_select"
      >
        Sort by
      </label>
      <select
        class="CustomSelect "
        id="searchForm_select"
        role="listbox"
      >
        <option
          selected=""
          value="Relevance"
        >
          Relevance
        </option>
        <option
          value="Newest"
        >
          Newest
        </option>
      </select>
    </div>
    <div
      class="SearchForm_button_container"
    >
      <button
        class="CustomButton "
        id="searchForm_submit_button"
        role="button"
      >
        Search
      </button>
    </div>
  </form>
  <div
    class="BooksTable_table"
  >
    <div
      class="BookCard"
    >
      <a
        class="BookCard_link_wrapper"
        href="/z7D2DwAAQBAJ"
        role="link"
      >
        <div
          class="BookCard_container"
        >
          <img
            alt="The Modern JavaScript Basics Tutorial by Be Sure Academy volume cover"
            class="BookCard_book_cover_image"
            src="http://books.google.com/books/content?id=z7D2DwAAQBAJ&printsec=frontcover&img=1&zoom=5&edge=curl&source=gbs_api"
          />
          <div
            class="BookCard_text_info_container"
          >
            <p
              class="BookCard_title"
            >
              The Modern JavaScript Basics Tutorial
            </p>
            <div>
              <p
                class="BookCard_authors"
              >
                Be Sure Academy
              </p>
              <p
                class="BookCard_categories"
              >
                Computers
              </p>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
  <div
    class="HomePage_loadMoreButton_container"
  >
    <button
      class="CustomButton "
      role="button"
    >
      Load more
    </button>
  </div>
</div>
`;
